### PR TITLE
ROCEv2 support

### PR DESCRIPTION
**Pull request by venkatmahalingam**
Add support for ROCEv2 feature.

### CODE REVIEW

File: doc/qos/SONiC_ROCEv2_support.md [hunk 1]
```diff
@@ -0,0 +1,488 @@
+# Feature Name
+RDMA (Remote Direct Memory Access) over Converged Ethernet (ROCEv2) support
```
Comment (prvattem) [inline doc/qos/SONiC_ROCEv2_support.md:2]: Please rephrase it to "Configuration Management support for ROCEv2" or "Management Framework support for ROCEv2"

File: doc/qos/SONiC_ROCEv2_support.md [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+
+# 1 Feature Overview
+
+ROCE provides the ability to transport Infiniband (IB) protocol traffic using RDMA over lossless converged Ethernet
+across storage or compute nodes.
+ROCEv2 stands for RDMA over Converged Ethernet version 2, this is an extension of ROCE protocol a.k.a routable ROCE,
+This overcomes the limitation of ROCEv1 bounded to a single broadcast domain (VLAN),
+ROCEv2 encapsulates an RDMA transport packet within the Ethernet/IPv4/UDP packet,
+ROVEv2 uses PFC to prevent buffer overflow.
+
```
Comment (prvattem) [inline doc/qos/SONiC_ROCEv2_support.md:60]: Please update this section with Management Framework capabilities that would be used in configuring the feature. Also provide info about REST/gNMI support.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+# 3 Design
+
+## 3.1 Overview
+
+Enhancing the management framework backend code and transformer methods to add support for QoS buffer.
+
+## 3.2 DB Changes
+
+### 3.2.1 CONFIG DB
+Existing QoS buffer tables will be used and a new field will be introduced in PORT table to avoid taking the buffer-mgr created profile based
```
Comment (prvattem) [inline doc/qos/SONiC_ROCEv2_support.md]: Is the PORT Table schema(with the new field) captured in this doc?

File: doc/qos/SONiC_ROCEv2_support.md (hunk 1 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md:2]: The scope is end-to-end ROCEv2 support along with mgmt-framework support, will rephrase accordingly.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 3 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md]: will update.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 2 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md:60]: This is feature overview, requirements section is updated with mgmt-fwk requirements.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+# 1 Feature Overview
+
+ROCE provides the ability to transport Infiniband (IB) protocol traffic using RDMA over lossless converged Ethernet
+across storage or compute nodes.
+ROCEv2 stands for RDMA over Converged Ethernet version 2, this is an extension of ROCE protocol a.k.a routable ROCE,
+This overcomes the limitation of ROCEv1 bounded to a single broadcast domain (VLAN),
+ROCEv2 encapsulates an RDMA transport packet within the Ethernet/IPv4/UDP packet,
+ROVEv2 uses PFC to prevent buffer overflow.
+
+## 1.1 Requirements
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md:61]: As of now only configuration requirements are mentioned here. Do we have any functional requirements for ROCEv2 feature validation? Request to provide any such details are available. What exactly we are going to validate as part of this feature testing? Please provide.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+
+
+# 1 Feature Overview
+
+ROCE provides the ability to transport Infiniband (IB) protocol traffic using RDMA over lossless converged Ethernet
+across storage or compute nodes.
+ROCEv2 stands for RDMA over Converged Ethernet version 2, this is an extension of ROCE protocol a.k.a routable ROCE,
+This overcomes the limitation of ROCEv1 bounded to a single broadcast domain (VLAN),
+ROCEv2 encapsulates an RDMA transport packet within the Ethernet/IPv4/UDP packet,
+ROVEv2 uses PFC to prevent buffer overflow.
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md]: Please change ROVEv2 to ROCEv2. Spell correction.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+| **Term**                 | **Meaning**                         |
+|--------------------------|-------------------------------------|
+| RDMA                     | Remote Direct Memory Access         |
+|          ROCEv2            |  RDMA over Converged Ethernet version 2           |
+|          PFC            |  Priority Flow Control|
+
+
+# 1 Feature Overview
+
+ROCE provides the ability to transport Infiniband (IB) protocol traffic using RDMA over lossless converged Ethernet
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md:54]: Is there any way to generate IB traffic for functional validation? Request to provide more details on generation of such traffic items.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+  .......
+  .......
+}                                                                                                                                                                        48        563,7         19%
+
+```
+The above tables are populated by default already in the SONiC system upon executing
+the 'config qos buffer reload' Click command, we are providing the following configuration options to fine tune the values
+and the get support will be provided to dump the buffer information.
+
+### 1.1.3 Functional Requirements
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md:169]: Does configuration of buffer pool and buffer profile is enough for validation or should we also test some functional behavior of those commands as part of the testing? Please confirm if only configuration part is enough for test plan preperation.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+
+Provide management framework support to
+- Configure buffer pool and buffer pool
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md]: I believe this buffer pool and buffer profile. Please change if my expectation is correct.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+- SONiC QOS buffer model on Redis DB schema
+- transformer functions to
+   * convert OpenConfig YANG model to SONiC YANG model for buffer related configurations
+
+### 1.2.3 SAI Overview
+
+# 2 Functionality
+
+## 2.1 Target Deployment Use Cases
+Manage/configure QoS buffer configurations via gNMI, REST and CLI interfaces
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md:206]: Do we have any functional use cases to be validated? Please confirm.

**Review by as943162 — COMMENTED** [review]
Please confirm if only configuration of the new commands validation is enough as part of this feature validation. Also provide if there is any way to validate IB traffic creation and validation of such traffic using our current feature.

Thanks in advance for your time in reviewing my comments and getting back.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+
+## 2.2 Functional Description
+Provide CLI, gNMI and REST support for QoS buffer onfigurations.
+
+## 2.3 Backend change to support new configurations
+Provide change in management framework and buffermgrd modules.
```
Comment (prvattem) [inline doc/qos/SONiC_ROCEv2_support.md:212]: if possible please capture the changes required in buffermgrd .

Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md:212]: sure

File: doc/qos/SONiC_ROCEv2_support.md (hunk 8 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md]: Yes, will change it.

**Comment by venkatmahalingam** [comment]
> Please confirm if only configuration of the new commands validation is enough as part of this feature validation. Also provide if there is any way to validate IB traffic creation and validation of such traffic using our current feature.
> 
> Thanks in advance for your time in reviewing my comments and getting back.

As part of the feature testing, we're simulating traffic congestion using IXIA to test PFC, I think, system test scope is to verify the lossless traffic from server/storage.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 5 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md]: ok

File: doc/qos/SONiC_ROCEv2_support.md (hunk 4 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md:61]: As part of the feature testing, we're simulating traffic congestion using IXIA to test PFC, I think, system test scope is to verify the lossless traffic from server/storage.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 6 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md:54]: As part of the feature testing, we're simulating traffic congestion using IXIA to test PFC, I think, system test scope is to verify the lossless IB traffic from server/storage.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 9 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md:206]: Please see above response.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 4 above)
Comment (as943162) [reply doc/qos/SONiC_ROCEv2_support.md:61]: Do you mean we need to generate L3( IPv4/IPv6 ) traffic using IXIA ports to create congestion and that congestion will be managed by PFC? If Yes, please confirm what is the role of ROCEv2 in that specific case. Thank you.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 11]
```diff
@@ -0,0 +1,654 @@
...
+All commands are executed in `configuration-view`:
+```
+sonic# configure terminal
+sonic(config)#
+```
+
+##### 3.6.2.1.1 Configure buffer pool
+Configure ingress and egress type buffer pools.
+```
+sonic(config)# qos buffer-pool <name> <shared-buffer-size-in-bytes> { [type ingress shared-headroom-size <xoff (included in the shared-buffer-size-in-bytes)>]  | [type egress ]}
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md]: I believe the buffer size has different values for each chipset. Please provide how do identify and configure buffer-size for different hardware/chipset dynamically.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 4 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md:61]: It is actually testing the PFC but the payload could be anything, system test is expected to test the IB traffic from server.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 7 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md:169]: We should verify the ROCEv2 traffic in the system test but the functional tests can be limited to test the PFC behaviour from IXIA.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 11 above)
Comment (venkatmahalingam) [reply doc/qos/SONiC_ROCEv2_support.md]: We should document the buffer size for each platform in the user guide.

File: doc/qos/SONiC_ROCEv2_support.md (hunk 4 above)
Comment (as943162) [reply doc/qos/SONiC_ROCEv2_support.md:61]: What exactly IB traffic means? How do we generate IB traffic? What kind of servers to be used to generate such traffic? Please clarify.

File: doc/qos/SONiC_ROCEv2_support.md [hunk 12]
```diff
@@ -0,0 +1,696 @@
...
+#### 3.6.2.1 Configuration Commands
+All commands are executed in `configuration-view`:
+```
+sonic# configure terminal
+sonic(config)#
+```
+##### 3.6.2.1.1 Initialize the buffer based on system defaults
+Initialize the buffer defaults based on platform specific values (ingress/ingress buffer pools size, buffer profile, priority-group, queue..etc)
+```
+sonic(config)# qos buffer init
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md:413]: Will it create a default lossless profile if the command executed? Also please let us know if default lossless profile is macro?

File: doc/qos/SONiC_ROCEv2_support.md [hunk 13]
```diff
@@ -0,0 +1,696 @@
...
+Clear default buffer initialization.
+```
+sonic(config)# no qos buffer init
+
+```
+##### 3.6.2.1.3 Configure buffer pool
+Configure shared head room size for the ingress buffer pool (fixed name - ingress_lossless_pool), other settings such as pool size and egress pool are automatically created during buffer init.
+Please refer user guide for platform specific defaults and based on that use these CLIs to fine-tune the numbers for the use-case.
+```
+sonic(config)# qos buffer pool ingress_lossless_pool shared-headroom-size shared-headroom-size <xoff>
```
Comment (as943162) [inline doc/qos/SONiC_ROCEv2_support.md:427]: There is a difference in the config command and the no form of this command. Config command doesnt contain '-' and no form of the command contains '-' between 'buffer' and 'pool'. What is the correct command syntax 'buffer-pool' or 'buffer pool'? Please confirm.

**Review by as943162 — COMMENTED** [review]
Please confirm the correct command 'buffer-pool' or 'buffer pool'.

Thanks in advance.

**Pull request merged**